File: src/content/parseVocabulary.ts
import { VocabularyEntry } from "./types";

/**
 * Parses a vocabulary file laid out as repeating groups of lines:
 *   German word
 *   Russian translation
 *   [pronunciation]   (optional — only consumed if present)
 *
 * Resilient to entries missing a pronunciation line, so it keeps working if a
 * future lesson's словарь file doesn't include one for every word.
 */
export function parseVocabulary(raw: string, lessonId: string): VocabularyEntry[] {
  const lines = raw
    .split(/\r\n|\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  const entries: VocabularyEntry[] = [];
  let i = 0;
  let index = 0;

  while (i < lines.length) {
    const german = lines[i++];
    if (i >= lines.length) break; // dangling word with no translation — drop it
    const translation = lines[i++];

    let pronunciation: string | undefined;
    const next = lines[i];
    if (next && /^\[.*\]$/.test(next)) {
      pronunciation = next.slice(1, -1).trim();
      i++;
    }

    entries.push({
      id: `${lessonId}-vocab-${index++}`,
      german,
      translation,
      pronunciation,
    });
  }

  return entries;
}

File: backend/app/legacy_parser/parse_vocabulary.py
"""Exact port of src/content/parseVocabulary.ts — see text_utils.py's module
docstring for why bit-for-bit fidelity matters here."""

import regex

_BRACKETED_LINE = regex.compile(r"^\[.*\]$")


def parse_vocabulary(raw: str, lesson_id: str) -> list[dict]:
    lines = [line.strip() for line in regex.split(r"\r\n|\n", raw)]
    lines = [line for line in lines if line]

    entries: list[dict] = []
    i = 0
    index = 0

    while i < len(lines):
        german = lines[i]
        i += 1
        if i >= len(lines):
            break  # dangling word with no translation — drop it
        translation = lines[i]
        i += 1

        pronunciation = None
        if i < len(lines) and _BRACKETED_LINE.match(lines[i]):
            pronunciation = lines[i][1:-1].strip()
            i += 1

        entries.append(
            {
                "id": f"{lesson_id}-vocab-{index}",
                "german": german,
                "translation": translation,
                "pronunciation": pronunciation,
            }
        )
        index += 1

    return entries
